dotnet.md

```diff
@@ -0,0 +1,95 @@
+7. Namespace imports should be specified at the top of the file, *outside* of `namespace` declarations.
+   - Namespaces should be sorted alphabetically, with the exception of `System.*` namespaces, which are to be placed on top of all others.
+8. Avoid more than one empty line at any time. For example, do not have two blank lines between members of a type.
+9. Avoid spurious free spaces. For example avoid `if (someVar == 0)...`, where the dots mark the spurious free spaces.
+   Consider enabling "View White Space (Ctrl+R, Ctrl+W)" or "Edit -> Advanced -> View White Space" if using Visual Studio to aid detection.
+10. If a file happens to differ in style from these guidelines (e.g. private members are named `m_member` rather than `_member`), the existing style in that file takes precedence.
+11. We only use `var` when it's obvious what the variable type is (e.g. `var stream = new FileStream(...)` not `var stream = OpenStandardInput()`).
+12. We use language keywords instead of BCL types (e.g. `int, string, float` instead of `Int32, String, Single`, etc) for both type references as well as method calls (e.g. `int.Parse` instead of `Int32.Parse`). See issue [391](https://github.com/dotnet/corefx/issues/391) for examples.
+13. We use PascalCasing to name all our constant local variables and fields. The only exception is for interop code where the constant value should exactly match the name and value of the code you are calling via interop.
+14. We use ```nameof(...)``` instead of ```"..."``` whenever possible and relevant.
+15. Fields should be specified at the top within type declarations.
+16. When including non-ASCII characters in the source code use Unicode escape sequences (\uXXXX) instead of literal characters. Literal non-ASCII characters occasionally get garbled by a tool or editor.
+17. When using labels (for goto), indent the label one less than the current indentation.
+18. When using a single-statement if, we follow these conventions:
+    - Never use single-line form (for example: `if (source == null) throw new ArgumentNullException("source");`)
+    - Using braces is always accepted, and required if any block of an `if`/`else if`/.../`else` compound statement uses braces or if a single statement body spans multiple lines.
+    - Braces may be omitted only if the body of *every* block associated with an `if`/`else if`/.../`else` compound statement is placed on a single line.
+19. String literals that will be user-facing should reference a [resource file](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization?view=aspnetcore-3.1#resource-files) entry. 
+    Resource files should be named using [this convention](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization?view=aspnetcore-3.1#resource-file-naming).
+	This applies to all submissions, and not just ASP.NET.
```

Comment:
Most of this is taken directly from the dotnet coreclr team's guidelines. Items 1 and 19 are my additions.

Here are my thoughts on changing the above (almost all related to minimizing typing):

## 4 and 5
> 4. We use `_camelCase` for internal and private fields and use `readonly` where possible. 
>    - Prefix internal and private instance fields with `_`, static fields with `s_` and thread static fields with `t_`. 
>    - When used on static fields, `readonly` should come after `static` (e.g. `static readonly` not `readonly static`).  
>    - Public fields should be used sparingly and should use PascalCasing with no prefix when used.
> 5. We avoid `this.` unless absolutely necessary. 

I disagree with these because applying the rules to test projects will either make writing test significantly more verbose or lead to violating 14 and being unable to use built-in refactoring tools to easily rename without breaking tests.

Instead, I would amend these to be:

> 4. We use `camelCase` for internal and private fields and use `readonly` where possible. 
>    - Do not prefix internal and private instance field names.
>    - When used on static fields, `readonly` should come after `static` (e.g. `static readonly` not `readonly static`).  
>    - Public fields should be used sparingly and should use PascalCasing with no prefix when used.
> 5. We avoid superfluous use of `this.`.

## 6

> 6. We always specify the visibility, even if it's the default (e.g. `private string _foo` not `string _foo`). 

I strongly prefer to **not** specify the default visibility to reduce typing. If this can be automatically applied during cleanup by `.editorconfig`, then I would not mind keeping the first sentence of this rule.

## 7
> Namespaces should be sorted alphabetically, with the exception of `System.*` namespaces, which are to be placed on top of all others.

That's probably useful for the coreclr team. I would prefer this rule simplly be "Namespaces should be sorted alphabetically." This is already automatically applied by `.editorconfig` during cleanup.

## 11
> 11. We only use `var` when it's obvious what the variable type is (e.g. `var stream = new FileStream(...)` not `var stream = OpenStandardInput()`).

I tried this rule in `.editorconfig` previously for a few months and absolutely _hated_ it. Arrangements in test code become verbose affairs whenever you mock a generic type (often). A series of initial assignments can result in large swaths of type declarations and variable names having a ragged left edge instead of all starting on the same column.

Since the IDE can provide this information through context, the primary benefit of this rule is for PR code reviews. I am inclined to ditch this rule and only request the type be spelled out where it is difficult for reviewers to infer from usage. I'm curious about others' opinions.

## 18

> 18. When using a single-statement if, we follow these conventions:
>     - Never use single-line form (for example: `if (source == null) throw new ArgumentNullException("source");`)
>     - Using braces is always accepted, and required if any block of an `if`/`else if`/.../`else` compound statement uses braces or if a single statement body spans multiple lines.
>     - Braces may be omitted only if the body of *every* block associated with an `if`/`else if`/.../`else` compound statement is placed on a single line.

I have a strong dislike for the "Never use single-line form", particularly for guard clauses like the example. If a single-statement `if` can easily fit on one line, I have always preferred to use a single line.

I would prefer to replace the first point with:
>     - Avoid single-line form except for guard clauses (for example: `if (source == null) throw new ArgumentNullException("source");`)

Reply:
Self-reminder to update the line reference for item 2 since the item it references has been renumbered.